---
title: "Cleaning genotype data for</br>diversity outbred mice"
author: "[Karl Broman](https://kbroman.org)"
date: 'Biostatistics & Medical Informatics<br/>University of Wisconsin&ndash;Madison<br/><a href="https://kbroman.org"><code>kbroman.org</code></a><br/><a href="https://github.com/kbroman"><code>github.com/kbroman</code></a><br/><a href="https://twitter.com/kwbroman"><code>@kwbroman</code></a>'
output:
    xaringan::moon_reader:
        lib_dir: libs
        nature:
            highlightStyle: github
            highlightLines: true
            countIncrementalSlides: false
            slideNumberFormat: '%current%'
            ratio: "16:9"
        css: [default, 'kbroman.css']
        chakra: libs/remark-latest.min.js
---

```{r setup, include=FALSE}
ratio <- "16:9"
fig_height <- ifelse(ratio=='4:3', 9, 7.5)
fig_width <- 15
curves_height <- ifelse(ratio=='4:3', 580, 540)
curves_width <- ifelse(ratio=='4:3', 800, 1080)
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(dev="svg", echo=FALSE, results="hide",
                      fig.height=fig_height,
                      fig.width=fig_width)
xaringan::summon_remark() # download latest version, locally
suppressMessages(library(fst))
library(qtl2)
library(broman)
library(jpeg)
suppressMessages(library(igraph))
iArrows <- igraph:::igraph.Arrows
library(lineup2)
library(knitr)
library(kableExtra)
```

## Heterogeneous stock

```{r dofig}
source("R/hs_fig.R")
```

---

## Diversity outbred mouse data

- Collaboration with Alan Attie, Gary Churchill, Brian Yandell,
  Josh Coon, Federico Rey, and many others

- 500 DO mice

- GigaMUGA SNP arrays

- RNA-seq data on pancreatic islets

- Microbiome data (16S and shotgun sequencing)

- protein and lipid measurements by mass spec


---


## Principles

What might have gone wrong?

How could it be revealed?



---

## Possible problems

- Sample duplicates

- Sample mix-ups

- Bad samples

- Bad markers

- Genotyping errors in founders


---

## Missing data per sample

```{r load_data}
x <- read.fst("Data/attieDO_chrXint.fst")
y <- read.fst("Data/attieDO_chrYint.fst")
do <- readRDS("Data/attieDO_v0.rds")

# for samples in replicate, figure out which one to keep
tab <- table(do$covar$mouse)
dup_mice <- names(tab[tab>1])
dup_mice <- sprintf("DO%03d", as.numeric(sub("DO", "", dup_mice)))
nm <- n_missing(do, summary="prop")
nm_before <- nm[dup_mice]
nm_after <- nm[paste0(dup_mice, "_b")]
keep <- names(nm_before)[nm_before < nm_after]

# samples to omit
omit <- dup_mice
omit[omit %in% keep] <- paste0(omit[omit %in% keep], "_b")

x <- x[,is.na(match(names(x), omit))]
y <- y[,is.na(match(names(y), omit))]
ids <- ind_ids(do)
do <- do[ids[!(ids %in% omit)],]
```

```{r plot_missing}
prop_missing <- sort(n_missing(do, "ind", "prop"))*100
par(mar=c(5.1, 4.1, 1.1, 1.1), cex=1.5, cex.axis=1.3, cex.lab=1.3)
grayplot(prop_missing, ylab="Percent missing genotypes",
         xlim=c(0, 502), xaxs="i", mgp.x=c(1.9,0.3,0),
         ylim=c(0, 70), yaxs="i")
index <- seq_along(prop_missing)
wh <- which(prop_missing > 10)
lab <- sub("_b$", "", names(prop_missing))
adj <- c(-0.5, +0.5, -1, +1, 0, 0, 0)
text(index[wh]-3, prop_missing[wh]+adj, lab[wh], adj=c(1, 0.5))
```

---

## Sex swaps


```{r yint_vs_xint_all_probes}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
xm <- colMeans(x[,-1], na.rm=TRUE)
ym <- colMeans(y[,-1], na.rm=TRUE)
purple_green <- brocolors("web")[c("purple", "green")]
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, "All markers", adj=1, cex=1.8)
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
```

---

## Sex swaps

```{r yint_vs_xint_selected_probes}
file <- "R/_cache/xymarkers.RData"
if(file.exists(file)) {
    load(file)
} else {
    xp <- apply(x[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    yp <- apply(y[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    xmar <- x[xp > -log10(0.05/length(xp)),1]
    ymar <- y[yp > -log10(0.05/length(yp)),1]
    save(xmar, ymar, file=file)
}
xm <- colMeans(x[x[,1] %in% xmar, -1], na.rm=TRUE)
ym <- colMeans(y[y[,1] %in% ymar, -1], na.rm=TRUE)
prop_missing <- n_missing(do, "ind", "prop")

yint_vs_xint_selected <-
function(labels=FALSE) {
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, adj=1, cex=1.8, "Selected markers")
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
if(!labels) return()
to_label <- c(names(sort(xm)[1:2]), # really low X intensity
              names(xm)[xm < 0.4 & ym < 0.1], # low X and Y
              names(xm)[do$is_female & ym > 0.2]) # female but high Y intensity
label <- sub("_b$", "", to_label)
yadd <- 0.02*c(1,1,1,1,-1.7,1)
xadj <- c(0.5, 0.5, 1, 0, 0.5, 0.5)
for(i in seq_along(to_label))
    text(xm[to_label[i]], ym[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
}
yint_vs_xint_selected(FALSE)
```

---

## Sex swaps

```{r yint_vs_xint_selected_probes_labeled}
yint_vs_xint_selected(TRUE)
```


---

## Sex swaps

```{r het_vs_xint}
prop_het <- rowSums(do$geno$X == 2)/rowSums(do$geno$X != 0)

het_vs_xint <- function(labels=FALSE) {
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, prop_het, bg=purple_green[do$is_female + 1],
         ylab="proportion het on X",
         xlab="average X chr intensity")
on_top <- (do$is_female & xm < 0.4)
points(xm[on_top], prop_het[on_top], pch=21, bg=purple_green[2])
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
if(!labels) return()
to_label <- c(names(xm)[xm < 0.4 & prop_het > 0.2],
              names(xm)[do$is_female & xm > 0.35 & xm < 0.4])
xadd <- 0.003*c(1,1,1,1,-2.7)
yadd <- 0.035*c(0,0,0,0,1)
xadj <- c(0,0,0,0,0.5)
label <- sub("_b$", "", to_label)
for(i in seq_along(to_label))
    text(xm[to_label[i]]+xadd[i], prop_het[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
}
het_vs_xint(FALSE)
```


---

## Sex swaps

```{r het_vs_xint_labeled}
het_vs_xint(TRUE)
```

---

## Sample duplicates

```{r prop_match}
file <- "R/_cache/compare_geno.rds"
if(file.exists(file)) {
    cg <- readRDS(file)
} else {
    cg <- compare_geno(do, cores=0)
    saveRDS(cg, file)
}
par(mar=c(6.1, 1.1, 1.1, 1.1), cex=1.7, cex.lab=1.3, cex.axis=1.3)
cg_upper <- cg[upper.tri(cg)]
hist(cg_upper, breaks=seq(0, 1, length=201),
     xlab="Proportion matching genotypes", ylab="", yaxt="n", main="")
rug(cg_upper, col="violetred")
```

---

## Sample duplicates

```{r prop_match_labeled}
par(mar=c(6.1, 1.1, 1.1, 1.1), cex=1.7, cex.lab=1.3, cex.axis=1.3)
hist(cg_upper, breaks=seq(0, 1, length=201),
     xlab="Proportion matching genotypes", ylab="", yaxt="n", main="")
rug(cg_upper, col="violetred")
textpos <- median(cg_upper[cg_upper < 0.35])
text(textpos, 4000, "DO306 / DO308", col="darkslateblue")
arrows(textpos, 3000, textpos, 1000, length=0.15, col="violetred", lwd=2)
```

---

## Sample mix-ups

```{r mixup_illustration_1}
source("R/gve_mixup_scheme.R")
gve_scheme(1)
```

---

## Sample mix-ups

```{r mixup_illustration_2}
gve_scheme(2)
```

---

## RNA-seq sample mix-ups: distance matrix

```{r gve_dist_image, dev="png", dpi=600}
load("Data/rnaseq_mixups_results.RData")
par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)
image(1:ncol(d), 1:nrow(d), t(d),
      col=gray((0:256)/256), las=1,
      xlab="DNA sample", ylab="RNA sample")
```

---

## RNA-seq sample mix-ups: minimum vs self distance

```{r gve_best_vs_self_unlabeled}
gve_best_vs_self <-
    function(labels=FALSE)
{
    selfd <- get_self(d)
    bestd <- get_best(d)
    par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)
    grayplot(selfd, bestd, xlab="distance to self", ylab="minimum distance")
    if(!labels) return()

    wh <- which(selfd > 0.8)
    xadd <- yadd <- setNames(rep(0, length(wh)), names(wh))
    yadd["DO340"] <- +0.02
    yadd["DO397"] <- -0.02
    yadd["DO306"] <- -0.025
    xadd["DO306"] <- +0.05
    yadd["DO268"] <- +0.02
    yadd["DO310"] <- -0.025
    xadd["DO310"] <- +0.05
    yadd["DO360"] <- -0.01
    yadd["DO370"] <- +0.005
    for(i in seq_along(wh)) {
        text(selfd[wh[i]]-0.008+xadd[i], bestd[wh[i]]+yadd[i], names(wh)[i], adj=c(1, 0.5))
    }
}
gve_best_vs_self(FALSE)
```

---

## RNA-seq sample mix-ups: minimum vs self distance

```{r gve_best_vs_self_labeled}
gve_best_vs_self(TRUE)
```


---

## RNA-seq sample mix-ups

```{r rnaseq_problems_detail}
label_sep <- 10
par(mfcol=c(2,3), mar=c(3.1,5.1,2.1,1.1), cex=1.3)
for(ind in paste0("DO", c(268, 269, 309, 310, 360, 370))) {
    this_d <- sort(d[,ind])
    grayplot(seq_along(this_d), this_d,
             xlab="", xaxs="i", xlim=c(-4.5, nrow(d)+0.5),
             ylab="distance", main=ind, ylim=range(d),
             xat=c(100, 200, 300), vlines=c(100, 200, 300))
    text(1+label_sep, this_d[1], names(this_d)[1], adj=0)
    if(ind=="DO309") text(2+label_sep, this_d[2], names(this_d)[2], adj=0)
}
```

---

## Microbiome data


```{r microbiome_illustration}
par(mar=rep(0, 4), bty="n")
plot(0, 0, type="n", xlim=c(0,fig_width*10), ylim=c(0, fig_height*10), xaxs="i", yaxs="i",
     xaxt="n", yaxt="n", xlab="", ylab="", pty="s")
mouse_image <- jpeg::readJPEG("Data/mouse_emoji.jpg")
poop_image <- jpeg::readJPEG("Data/poop_emoji.jpg")
dna_image <- jpeg::readJPEG("Data/dna_emoji.jpg")
xpos <- 45;ypos <- 70
width <- 16;height <- 16
rasterImage(mouse_image, xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 80;ypos<-60
rasterImage(poop_image,  xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 110;ypos<-45
rasterImage(dna_image,   xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 25;ypos<-40
rasterImage(dna_image,   xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)

arrow_color <- "violetred"
arrow_lwd <- 4
arrowhead <- 0.7

# mouse to DNA on left
iArrows(35, 66, 25, 50, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=-0.3, width=1, size=arrowhead)
# mouse to poop on right
iArrows(55, 67, 70, 62, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.1, width=1, size=arrowhead)
# poop to DNA on right
iArrows(90, 56, 105, 48, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.1, width=1, size=arrowhead)
# DNA to "snps" on left
iArrows(24, 32, 24, 20, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0, width=1, size=arrowhead)
text(25, 15, "SNPs", cex=3, col="darkslateblue")
# DNA to shotgun sequencing on right
iArrows(115, 40, 120, 33, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.2, width=1, size=arrowhead)
set.seed(31168898)
start <- 95; end <- 140
n_reads <- 200
read_length <- (end-start) * runif(n_reads, 0.06, 0.075)
read_start <- runif(n_reads, start, end-read_length)
read_end <- read_start + read_length
y <- qtl2:::arrange_genes(read_start, read_end)
y <- y[1:100]
top <- 30; bottom <- 10
y <- max(y) - y
y <- (y-min(y))*(top-bottom)/(max(y)-min(y)) + bottom
segments(read_start, y, read_end, y, lwd=2, col="slateblue")
```

---

## Sample mix-ups: Microbiome data


- Impute genotypes at all SNPs in DNA samples

- Map microbiome reads to mouse genome; find reads overlapping a SNP

- For each pair of samples (DNA + microbiome):

  - Focus on reads that overlap a SNP where that DNA sample is
    homozygous

  - Distance = proportion of reads where SNP allele doesn't match DNA
    sample's genotype


---

## Microbiome DO361 vs DNA DO361

```{r do361_do361, results="asis"}
microbiome <- readRDS("Data/microbiome_mixups_dist.rds")
tab <- table("Microbiome DO361"=c("A", "B"), "DNA DO361"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO361", "DO361"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO361", "DO361"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO361", "DO361"]
tab[2,2] <- microbiome$k3["DO361", "DO361"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```




---

## Microbiome DO360 vs DNA DO360

```{r do360_do360, results="asis"}
tab <- table("Microbiome DO360"=c("A", "B"), "DNA DO360"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO360", "DO360"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO360", "DO360"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO360", "DO360"]
tab[2,2] <- microbiome$k3["DO360", "DO360"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```

---

## Microbiome DO360 vs DNA DO370

```{r do370_do360, results="asis"}
tab <- table("Microbiome DO360"=c("A", "B"), "DNA DO370"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO370", "DO360"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO370", "DO360"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO370", "DO360"]
tab[2,2] <- microbiome$k3["DO370", "DO360"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```

---

## Microbiome mix-ups

```{r microbiome_best_vs_self}
microbiome_best_vs_self <-
    function(labels=0)
{
    par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)
    grayplot(microbiome$self, microbiome$best, xlab="distance to self", ylab="minimum distance")
    if(labels==1) { # mix-ups

    }
    if(labels==2) { # bad samples

    }
    if(labels==3) { # duplicates

    }
    if(labels==3) { # mixtures

    }
}
microbiome_best_vs_self(0)
```



---

## Sample quality

- Proportion missing data

- Genotype frequencies

- No. crossovers

- No. apparent genotyping errors

---

## Marker quality

- Proportion missing data

- Genotype frequencies

- No. apparent genotyping errors


---

## Founder genotyping errors



---

## Summary

- Quality of results depends on quality of data

- Think about what might have gone wrong, and how it might be revealed

- Consider the order in apply procedures
