---
title: "Cleaning genotype data for</br>diversity outbred mice"
author: "[Karl Broman](https://kbroman.org)"
date: 'Biostatistics & Medical Informatics<br/>University of Wisconsin&ndash;Madison<br/><br/><a href="https://kbroman.org"><code>kbroman.org</code></a><br/><a href="https://github.com/kbroman"><code>github.com/kbroman</code></a><br/><a href="https://twitter.com/kwbroman"><code>@kwbroman</code></a><br/>Slides: <a href="https://bit.ly/2018ctc"><code>bit.ly/2018ctc</code></a>'
output:
    xaringan::moon_reader:
        lib_dir: libs
        nature:
            highlightStyle: github
            highlightLines: true
            countIncrementalSlides: false
            slideNumberFormat: '%current%'
            ratio: "16:9"
        css: [default, 'kbroman.css']
        chakra: libs/remark-latest.min.js
---

```{r setup, include=FALSE}
ratio <- "16:9"
fig_height <- ifelse(ratio=='4:3', 9, 7.5)
fig_width <- 15
curves_height <- ifelse(ratio=='4:3', 580, 540)
curves_width <- ifelse(ratio=='4:3', 800, 1080)
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(dev="svg", echo=FALSE, results="hide",
                      fig.height=fig_height,
                      fig.width=fig_width)
xaringan::summon_remark() # download latest version, locally
suppressMessages(library(fst))
library(qtl2)
library(qtl2fst)
library(broman)
library(jpeg)
suppressMessages(library(igraph))
iArrows <- igraph:::igraph.Arrows
library(lineup2)
library(knitr)
library(kableExtra)
```

# Heterogeneous stock

```{r dofig}
source("R/hs_fig.R")
```

---

# Diversity outbred mouse data

- 500 DO mice

- GigaMUGA SNP arrays (114k SNPs)

- RNA-seq data on pancreatic islets

- Microbiome data (16S and shotgun sequencing)

- protein and lipid measurements by mass spec

- Collaboration with Alan Attie, Gary Churchill, Brian Yandell,
  Josh Coon, Federico Rey, and many others

---


# Principles

What might have gone wrong?

How could it be revealed?

--

Also, just make a bunch of graphs.

--

If you see something weird, try to figure it out.



---

# Possible problems

- Sample duplicates

- Sample mix-ups

- Bad samples

- Bad markers

- Genotyping errors in founders


---
class: inverse, middle, center

# What to look at first?


---

# Missing data per sample

```{r load_data}
x <- read.fst("Data/attieDO_chrXint.fst")
y <- read.fst("Data/attieDO_chrYint.fst")
do <- readRDS("Data/attieDO_v0.rds")

# for samples in replicate, figure out which one to keep
tab <- table(do$covar$mouse)
dup_mice <- names(tab[tab>1])
dup_mice <- sprintf("DO%03d", as.numeric(sub("DO", "", dup_mice)))
nm <- n_missing(do, summary="prop")
nm_before <- nm[dup_mice]
nm_after <- nm[paste0(dup_mice, "_b")]
keep <- names(nm_before)[nm_before < nm_after]

# samples to omit
omit <- dup_mice
omit[omit %in% keep] <- paste0(omit[omit %in% keep], "_b")

x <- x[,is.na(match(names(x), omit))]
y <- y[,is.na(match(names(y), omit))]
ids <- ind_ids(do)
do <- do[ids[!(ids %in% omit)],]

# clean up the names
names(x) <- sub("_b$", "", names(x))
names(y) <- sub("_b$", "", names(y))
for(i in seq_along(do$geno)) {
    rownames(do$geno[[i]]) <- sub("_b$", "", rownames(do$geno[[i]]))
}
rownames(do$covar) <- sub("_b$", "", rownames(do$covar))
rownames(do$cross_info) <- sub("_b$", "", rownames(do$cross_info))
names(do$is_female) <- sub("_b$", "", names(do$is_female))

# reorder by mouse number
x <- x[, c(1,1+order(as.numeric(sub("DO", "", names(x)[-1]))))]
y <- y[, c(1,1+order(as.numeric(sub("DO", "", names(y)[-1]))))]
do <- do[order(as.numeric(sub("DO", "", ind_ids(do)))),]
```

```{r plot_missing}
prop_missing <- n_missing(do, "ind", "prop")*100
plot_missing_byind <- function(label=FALSE)
{
    mousenum <- as.numeric(sub("DO", "", names(prop_missing)))
    prop_missing <- prop_missing[order(mousenum)]

    par(mar=c(5.1, 4.1, 1.1, 1.1), cex=1.5, cex.axis=1.3, cex.lab=1.3)
    index <- setNames(seq_along(prop_missing), names(prop_missing))
    color <- qtl2::CCcolors[c(8,2,4,6,7)]
    wave <- as.numeric(do$covar[names(prop_missing),"wave"])

    grayplot(index, prop_missing, bg=color[wave],
             xlab="Mouse", ylab="Percent missing genotypes",
             xlim=c(0, 502), xaxs="i", mgp.x=c(1.9,0.3,0),
             ylim=c(0, 70), yaxs="i")

    if(!label) return()

    to_label <- names(prop_missing)[prop_missing > 10]
    xadj <- setNames(rep(+1, length(to_label)), to_label)
    xadj["DO357"] <- xadj["DO397"] <- -1

    for(i in to_label) {
        text(index[i]-3*xadj[i], prop_missing[i], i,
             adj=c(ifelse(xadj[i] > 0, 1, 0), 0.5))
    }
}
plot_missing_byind(FALSE)
```

---

# Missing data per sample

```{r missing_genotypes_labeled}
plot_missing_byind(TRUE)
```


---
class: inverse, middle, center

# Swapped sex labels

---

# Average SNP intensity on X and Y chr


```{r yint_vs_xint_all_probes}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
xm <- colMeans(x[,-1], na.rm=TRUE)
ym <- colMeans(y[,-1], na.rm=TRUE)
purple_green <- brocolors("web")[c("purple", "green")]
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, "All markers", adj=0, cex=1.8)
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
```

---

# Average SNP intensity on X and Y chr

```{r yint_vs_xint_selected_probes}
file <- "R/_cache/xymarkers.RData"
if(file.exists(file)) {
    load(file)
} else {
    xp <- apply(x[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    yp <- apply(y[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    xmar <- x[xp > -log10(0.05/length(xp)),1]
    ymar <- y[yp > -log10(0.05/length(yp)),1]
    save(xmar, ymar, file=file)
}
xm <- colMeans(x[x[,1] %in% xmar, -1], na.rm=TRUE)
ym <- colMeans(y[y[,1] %in% ymar, -1], na.rm=TRUE)
prop_missing <- n_missing(do, "ind", "prop")

yint_vs_xint_selected <-
function(labels=FALSE) {
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, adj=0, cex=1.8, "Selected markers")
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
if(!labels) return()
to_label <- c(names(sort(xm)[1:2]), # really low X intensity
              names(xm)[xm < 0.4 & ym < 0.1], # low X and Y
              names(xm)[do$is_female & ym > 0.2]) # female but high Y intensity
yadd <- setNames(rep(0.02, length(to_label)), to_label)
yadd["DO357"] <- yadd["DO357"] * -1.7
xadj <- setNames(rep(0.5, length(to_label)), to_label)
xadj["DO153"] <- 1
xadj["DO347"] <- 0
for(i in to_label)
    text(xm[i], ym[i]+yadd[i], i, adj=c(xadj[i], 0.5))
}
yint_vs_xint_selected(FALSE)
```

---

# Average SNP intensity on X and Y chr

```{r yint_vs_xint_selected_probes_labeled}
yint_vs_xint_selected(TRUE)
```


---

# Heterozygosity vs SNP intensity on X chr

```{r het_vs_xint}
prop_het <- rowSums(do$geno$X == 2)/rowSums(do$geno$X != 0)

het_vs_xint <- function(labels=FALSE) {
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, prop_het, bg=purple_green[do$is_female + 1],
         ylab="proportion het on X",
         xlab="average X chr intensity")
on_top <- (do$is_female & xm < 0.4)
points(xm[on_top], prop_het[on_top], pch=21, bg=purple_green[2])
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
if(!labels) return()
to_label <- c(names(xm)[xm < 0.4 & prop_het > 0.2],
              names(xm)[do$is_female & xm > 0.35 & xm < 0.4])
xadd <- 0.003*c(1,1,1,1,-2.7)
yadd <- 0.035*c(0,0,0,0,1)
xadj <- c(0,0,0,0,0.5)
for(i in seq_along(to_label))
    text(xm[to_label[i]]+xadd[i], prop_het[to_label[i]]+yadd[i], to_label[i],
         adj=c(xadj[i], 0.5))
}
het_vs_xint(FALSE)
```


---

# Heterozygosity vs SNP intensity on X chr

```{r het_vs_xint_labeled}
het_vs_xint(TRUE)
```

---
class: inverse, middle, center

# Sample duplicates

---

# Percent matching genotypes between pairs

```{r prop_match, dev="png", dpi=600}
file <- "R/_cache/compare_geno.rds"
if(file.exists(file)) {
    cg <- readRDS(file)
} else {
    cg <- compare_geno(do, cores=0)
    saveRDS(cg, file)
}
prop_match_fig <- function(label=FALSE)
{
    par(mar=c(6.1, 1.1, 1.1, 1.1), cex=1.7, cex.lab=1.3, cex.axis=1.3)
    cg_upper <- cg[upper.tri(cg)]*100
    hist(cg_upper, breaks=seq(0, 100, length=201),
         xlab="Percent matching genotypes", ylab="", yaxt="n", main="")
    rug(cg_upper, col="violetred")

    if(!label) return()

    textpos <- median(cg_upper[cg_upper < 35])
    text(textpos, 4000, "DO306 / DO308", col="darkslateblue")
    arrows(textpos, 3000, textpos, 1000, length=0.15, col="violetred", lwd=2)
}
prop_match_fig(FALSE)
```

---

# Percent matching genotypes between pairs

```{r prop_match_labeled, dev="png", dpi=600}
prop_match_fig(TRUE)
```

---
class: inverse, middle, center

# Sample mix-ups

---

# Sample mix-ups

```{r mixup_illustration_1}
source("R/gve_mixup_scheme.R")
gve_scheme(1)
```

---

# Sample mix-ups

```{r mixup_illustration_2}
gve_scheme(2)
```

---

# RNA-seq sample mix-ups: distance matrix

```{r gve_dist_image, dev="png", dpi=600}
load("Data/rnaseq_mixups_results.RData")
par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)
image(1:ncol(d), 1:nrow(d), t(d),
      col=gray((0:256)/256), las=1,
      xlab="DNA sample", ylab="RNA sample")
```

---

# RNA-seq sample mix-ups: minimum vs self distance

```{r gve_best_vs_self_unlabeled}
gve_best_vs_self <-
    function(labels=FALSE)
{
    selfd <- get_self(d)
    bestd <- get_best(d)
    par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)
    grayplot(selfd, bestd, xlab="distance to self", ylab="minimum distance")
    if(!labels) return()

    wh <- which(selfd > 0.8)
    xadd <- yadd <- setNames(rep(0, length(wh)), names(wh))
    yadd["DO340"] <- +0.02
    yadd["DO397"] <- -0.02
    yadd["DO306"] <- -0.025
    xadd["DO306"] <- +0.05
    yadd["DO268"] <- +0.02
    yadd["DO310"] <- -0.025
    xadd["DO310"] <- +0.05
    yadd["DO360"] <- -0.01
    yadd["DO370"] <- +0.005
    for(i in seq_along(wh)) {
        text(selfd[wh[i]]-0.008+xadd[i], bestd[wh[i]]+yadd[i], names(wh)[i], adj=c(1, 0.5))
    }
}
gve_best_vs_self(FALSE)
```

---

# RNA-seq sample mix-ups: minimum vs self distance

```{r gve_best_vs_self_labeled}
gve_best_vs_self(TRUE)
```


---

# RNA-seq sample mix-ups: detail

```{r rnaseq_problems_detail}
label_sep <- 10
par(mfcol=c(2,3), mar=c(3.1,5.1,2.1,1.1), cex=1.3)
for(ind in paste0("DO", c(268, 269, 309, 310, 360, 370))) {
    this_d <- d[,ind]
    this_d <- this_d[order(as.numeric(sub("DO", "", names(this_d))))]
    grayplot(seq_along(this_d), this_d,
             xlab="", xaxs="i", xlim=c(-4.5, nrow(d)+0.5),
             ylab="distance", main=ind, ylim=range(d),
             xat=c(100, 200, 300), vlines=c(100, 200, 300))
    text(which.min(this_d)-10, min(this_d), names(which.min(this_d)),
         adj=c(1, 0.5), cex=1.2)
}
```

---

# Microbiome data


```{r microbiome_illustration}
par(mar=rep(0, 4), bty="n")
plot(0, 0, type="n", xlim=c(0,fig_width*10), ylim=c(0, fig_height*10), xaxs="i", yaxs="i",
     xaxt="n", yaxt="n", xlab="", ylab="", pty="s")
mouse_image <- jpeg::readJPEG("Data/mouse_emoji.jpg")
poop_image <- jpeg::readJPEG("Data/poop_emoji.jpg")
dna_image <- jpeg::readJPEG("Data/dna_emoji.jpg")
xpos <- 45;ypos <- 70
width <- 16;height <- 16
rasterImage(mouse_image, xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 80;ypos<-60
rasterImage(poop_image,  xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 110;ypos<-45
rasterImage(dna_image,   xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)
xpos <- 25;ypos<-40
rasterImage(dna_image,   xpos-width/2,ypos-height/2,xpos+width/2,ypos+height/2)

arrow_color <- "violetred"
arrow_lwd <- 4
arrowhead <- 0.7

# mouse to DNA on left
iArrows(35, 66, 25, 50, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=-0.3, width=1, size=arrowhead)
# mouse to poop on right
iArrows(55, 67, 70, 62, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.1, width=1, size=arrowhead)
# poop to DNA on right
iArrows(90, 56, 105, 48, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.1, width=1, size=arrowhead)
# DNA to "snps" on left
iArrows(24, 32, 24, 20, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0, width=1, size=arrowhead)
text(25, 15, "SNPs", cex=3, col="darkslateblue")
# DNA to shotgun sequencing on right
iArrows(115, 40, 120, 33, h.lwd=arrow_lwd, sh.lwd=arrow_lwd, sh.col=arrow_color,
        curve=0.2, width=1, size=arrowhead)
set.seed(31168898)
start <- 95; end <- 140
n_reads <- 200
read_length <- (end-start) * runif(n_reads, 0.06, 0.075)
read_start <- runif(n_reads, start, end-read_length)
read_end <- read_start + read_length
y <- qtl2:::arrange_genes(read_start, read_end)
y <- y[1:100]
top <- 30; bottom <- 10
y <- max(y) - y
y <- (y-min(y))*(top-bottom)/(max(y)-min(y)) + bottom
segments(read_start, y, read_end, y, lwd=2, col="slateblue")
```

---

# Sample mix-ups: Microbiome data


- Impute genotypes at all SNPs in DNA samples

- Map microbiome reads to mouse genome; find reads overlapping a SNP

- For each pair of samples (DNA + microbiome):

  - Focus on reads that overlap a SNP where that DNA sample is
    homozygous

  - Distance = proportion of reads where SNP allele doesn't match DNA
    sample's genotype


---

# Microbiome DO361 vs DNA DO361

```{r do361_do361, results="asis"}
microbiome <- readRDS("Data/microbiome_mixups_dist.rds")
tab <- table("Microbiome DO361"=c("A", "B"), "DNA DO361"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO361", "DO361"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO361", "DO361"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO361", "DO361"]
tab[2,2] <- microbiome$k3["DO361", "DO361"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```




---

# Microbiome DO360 vs DNA DO360

```{r do360_do360, results="asis"}
tab <- table("Microbiome DO360"=c("A", "B"), "DNA DO360"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO360", "DO360"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO360", "DO360"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO360", "DO360"]
tab[2,2] <- microbiome$k3["DO360", "DO360"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```

---

# Microbiome DO360 vs DNA DO370

```{r do370_do360, results="asis"}
tab <- table("Microbiome DO360"=c("A", "B"), "DNA DO370"=c("AA", "BB"))
tab[1,1] <- microbiome$k1["DO370", "DO360"]
tab[2,1] <- (microbiome$n1 - microbiome$k1)["DO370", "DO360"]
tab[1,2] <- (microbiome$n3 - microbiome$k3)["DO370", "DO360"]
tab[2,2] <- microbiome$k3["DO370", "DO360"]
kable_styling(kable(add_commas(tab), align="rrr"),
              font_size=48, bootstrap_options="striped")
```

---

# Microbiome mix-ups: minimum vs self distance

```{r microbiome_best_vs_self}
microbiome_best_vs_self <- function(label=FALSE)
{
    par(mar=c(6.6,12.1,1.1, 12.1), cex=1.3, cex.axis=1.2, cex.lab=1.2)

    # color things
    bgcol <- setNames(rep("lightblue", length(microbiome$self)), names(microbiome$self))
    if(label) {
        mixups <- sprintf("DO%03d", c(360,370,53,54,83,85,88))
        dups <- sprintf("DO%03d", c(327,336))
        mixtures <- sprintf("DO%03d", c(329,340,343,344,346,354,359,362,   385,41))
        bad <- sprintf("DO%03d", c(357,397))
        bad_micro <- "DO174"
        weak <- c("DO205", "DO358")
        bgcol[mixups] <- "violetred"
        bgcol[bad] <- brocolors("web")["purple"]
        bgcol[mixtures] <- brocolors("web")["green"]
        bgcol[dups] <- "violetred"
        bgcol[bad_micro] <- brocolors("web")["purple"]
    }


    grayplot(microbiome$self, microbiome$best, bg=bgcol,
             xlab="distance to self", ylab="minimum distance",
             xlim=c(0, 0.22), ylim=c(0, 0.22), yaxs="i", xaxs="i")

    if(!label) return()

    par(cex=1.5)
    text(0.17, 0.145, "bad DNA", col=brocolors("web")["purple"])
    text(0.18, 0.025, "mix-ups", col="violetred")
    text(0.14, 0.09, "mixtures", col=brocolors("web")["green"], adj=c(1, 0.5))
    text(0.102, 0.048, "few reads", col=brocolors("web")["purple"], adj=c(1, 0.5))
}
microbiome_best_vs_self(FALSE)
```

---

# Microbiome mix-ups: minimum vs self distance

```{r microbiome_best_vs_self_labeled}
microbiome_best_vs_self(TRUE)
```


---
class: inverse, middle, center

# Sample quality

---

# Sample quality

- Proportion missing data

- SNP array intensities

- Genotype frequencies

- No. crossovers

- No. genotyping errors

---

# Missing data per sample

![](index_files/figure-html/missing_genotypes_labeled-1.svg)

---

# Array intensities

---

# Genotype frequencies, by individual

```{r genotype_freq}
# pull out genotype data
g <- do.call("cbind", do$geno[1:19])
fg <- do.call("cbind", do$founder_geno[1:19])

# omit markers with any missing founder genotypes
g <- g[,colSums(fg==0)==0]
fg <- fg[,colSums(fg==0)==0]

# 0 -> NA
g[g==0] <- NA

# recode so 1 = major allele
af_fg <- colSums(fg==1)
fg[,af_fg > 4] <- 4-fg[,af_fg > 4]
g[,af_fg > 4] <- 4-g[,af_fg > 4]
af_fg <- colSums(fg==1)

# genotype frequencies by marker class
af_g <- matrix(ncol=4, nrow=nrow(g))
dimnames(af_g) <- list(rownames(g), 1:4)
for(i in 1:4) {
    af_g[,i] <- rowMeans((g[,af_fg==i] - 1)/2, na.rm=TRUE)
}

prop_missing <- n_missing(do, "ind", "prop")*100

plot_geno_freq_byind <- function(subset=FALSE, density=FALSE)
{
    these_af <- af_g
    if(subset) {
        these_af <- af_g[prop_missing < 10, ]
    }

    par(mfrow=c(2,2), mar=c(5.1, 1.1, 2.1, 1.1), cex=1.5)
    for(i in 1:4) {
        if(subset) breaks <- seq(i/8-0.05, i/8+0.05, len=45) else breaks <- 80

        yli <- range(hist(1-these_af[,i], breaks=breaks, plot=FALSE)$density)

        if(density) {
            u <-par("usr")[1:2]
            u <- seq(min(breaks), max(breaks), len=251)
            dn <- dnorm(u, i/8, sqrt(i/8*(1-i/8) / sum(af_fg==i)))
            yli <- range(c(yli, dn))
        }

        hist(1-these_af[,i], breaks=breaks, ylim=yli,
             main=paste0("MAF = ", i, "/8"),
             yaxt="n", ylab="", xlab="Frequency of minor allele", prob=TRUE)
        arrows(i/8, par("usr")[3], i/8, 0, len=0.1, lwd=3, col="violetred")

        if(density) lines(u, dn, lwd=2, col="slateblue")
    }
}
plot_geno_freq_byind(FALSE)
```

---

# Genotype frequencies, by individual

```{r genotype_freq_subset}
plot_geno_freq_byind(TRUE)
```

---

# Genotype frequencies, by individual

```{r genotype_freq_subset_wdensity}
plot_geno_freq_byind(TRUE, TRUE)
```

---

# Genotype probabilities (one mouse on one chr)

```{r plot_genoprob, dev="png", dpi=600}
pr_fst <- readRDS("Data/GenoprobsFST/attieDO_v1_pr_fst.rds")
gmap <- readRDS("Data/attieDO_gmap.rds")
pmap <- readRDS("Data/attieDO_pmap.rds")
par(mar=c(5.1,5.1,2.1,0.1), cex=1.8, cex.axis=0.9, cex.lab=1.3, cex.main=1.3)
plot_genoprob(pr_fst, gmap, ind="DO361", chr=3, threshold=0.25,
              xlab="Chr 3 position (cM)", main="DO361", ylab="Genotype")
```

---

# Genome reconstruction (one mouse)

```{r plot_onegeno}
file <- "R/_cache/infg_nxo.RData"
if(file.exists(file)) {
    load(file)
} else {
    infg <- maxmarg(pr_fst, minprob=0.5, cores=0)
    infg_ph <<- guess_phase(do, infg, cores=0)
    n_xo <- count_xo(infg, cores=0)
    save(infg, infg_ph, n_xo, file=file)
}
par(mar=c(5.1,5.1,2.1,0.1), cex=1.8, cex.axis=1.1, cex.lab=1.3, cex.main=1.3)
plot_onegeno(infg_ph, gmap, ind="DO361", ylab="Position (cM)", main="DO361")
```

---

# Percent missing vs number of crossovers

```{r missing_v_nxo}
missing_v_nxo_fig <- function(label=FALSE, subset=FALSE)
{
    par(mar=c(5.1,5.1,1.1,1.1), cex=1.8, cex.axis=1.3, cex.lab=1.3, cex.main=1.3)
    prop_missing <- n_missing(do, "ind", "prop")*100
    tot_xo <- rowSums(n_xo)
    if(subset) {
        tot_xo <- tot_xo[prop_missing < 10]
        prop_missing <- prop_missing[prop_missing < 10]
    }
    grayplot(tot_xo, prop_missing, xlab="Number of crossovers", ylab="Percent missing data")
    if(!label) return()
    if(!subset) {
        to_lab <- names(prop_missing)[prop_missing >= 10]
    } else {
        to_lab <- names(tot_xo)[tot_xo == min(tot_xo) | tot_xo == max(tot_xo) | prop_missing > 5.9]
    }
    xadj <- setNames(rep(ifelse(subset, 5, 50), length(to_lab)), to_lab)
    yadj <- setNames(rep(0, length(to_lab)), to_lab)
    if(!subset) {
        xadj["DO340"] <- xadj["DO340"] * -1
        xadj["DO397"] <- xadj["DO397"] * -1
    } else {
        xadj["DO029"] <- 0; yadj["DO029"] <- 0.4
        xadj["DO524"] <- 0; yadj["DO524"] <- 0.4
    }
    for(i in seq_along(to_lab)) {
        text(tot_xo[to_lab[i]]+xadj[i], prop_missing[to_lab[i]]+yadj[i], to_lab[i],
             adj=c(ifelse(xadj[i] < 0, 1, ifelse(yadj[i] > 0, 0.5, 0)), 0.5))
    }
}
missing_v_nxo_fig(FALSE, FALSE)
```



---

# Percent missing vs number of crossovers

```{r missing_v_nxo_labeled}
missing_v_nxo_fig(TRUE, FALSE)
```

---

# Percent missing vs number of crossovers

```{r missing_v_nxo_subset}
missing_v_nxo_fig(FALSE, TRUE)
```

---

# Percent missing vs number of crossovers

```{r missing_v_nxo_subset_labeled}
missing_v_nxo_fig(TRUE, TRUE)
```

---

# No. crossovers by generation

```{r nxo_by_wave}
tot_xo <- rowSums(n_xo)
tot_xo <- tot_xo[tot_xo < 1000]
wave <- as.numeric(do$covar[names(tot_xo), "wave"])
mousenum <- as.numeric(sub("DO", "", names(tot_xo)))
o <- order(mousenum)
color <- qtl2::CCcolors[c(8,2,4,6,7)]
par(mar=c(5.1,5.1,2.1,1.1), cex=1.8, cex.axis=1.3, cex.lab=1.3, cex.main=1.3)
grayplot(seq_along(o), tot_xo[o], bg=color[wave[o]],
         xlab="Mouse number", ylab="No. crossovers", xaxs="i",
         xlim=c(-1.5,500.5))
gen <- tapply(do$covar$gen, do$covar$wave, "[", 1)
text(seq(50, 450, by=100), rep(617, 5), paste("gen", gen),
     xpd=TRUE, col=color)
```

---

# Estimated percent of genotyping errors

```{r genotype_errors}
file <- "R/_cache/n_err.RData"
if(file.exists(file)) {
    load(file)
} else {
    errlod_file <- "R/_cache/errlod.rds"
    if(file.exists(errlod_file)) {
        errlod <- readRDS(errlod_file)
    } else {
        errlod <- calc_errorlod(do, pr_fst, cores=0)
        saveRDS(errlod, errlod_file)
    }
    errlod <- do.call("cbind", errlod)
    nerr_byind <- rowSums(errlod > 4)
    nerr_bymar <- colSums(errlod > 4)
    save(nerr_byind, nerr_bymar, file=file)
}
plot_geno_err_byind <- function(subset=FALSE)
{
    par(mar=c(5.1,5.1,1.1,1.1), cex=1.5, cex.lab=1.3, cex.axis=1.3)
    yval <- nerr_byind/n_typed(do)*100
    if(subset) {
        prop_missing <- n_missing(do, "ind", "prop")*100
        yval <- yval[names(prop_missing)[prop_missing < 10]]
    }
    mousenum <- as.numeric(sub("DO", "", names(yval)))
    yval <- yval[order(mousenum)]
    xval <- setNames(seq_along(yval), names(yval))

    if(!subset) ylim <- c(0, 6.2)
    else ylim <- c(0, 0.27)

    grayplot(xval, yval,
             xlab="Mouse", ylab="Percent genotyping errors",
             xlim=c(-2.5, 502.5), xaxs="i", yaxs="i", ylim=ylim,
             mgp.y=c(3.1,0.3,0))

    # labels
    to_lab <- names(yval)[yval > ifelse(subset, 0.15, 0.25)]
    xadj <- setNames(rep(1, length(to_lab)), to_lab)
    if(subset) {
        xadj["DO135"] <- -1
    }
    for(i in to_lab) text(xval[i]+3*xadj[i], yval[i], i, adj=c(ifelse(xadj[i]<0,1,0), 0.5))
}
plot_geno_err_byind(FALSE)
```

---

# Estimated percent of genotyping errors

```{r genotype_errors_subset}
plot_geno_err_byind(TRUE)
```



---
class: inverse, middle, center

# Marker quality

---

# Marker quality

- Proportion missing data

- Genotype frequencies

- No. apparent genotyping errors

---

# SNP intensities at selected markers


---

# Proportion missing data

---

# Genotype frequencies

---

# Proportion of apparent errors

---

# Apparent errors, shown in intensity plots

---

# Area covered by spanning polygons


---
class: inverse, middle, center

# Founder genotyping errors

---

# Founder genotyping errors

- Apparent genotyping errors by inferred genotype


---

# Summary

- Quality of results depends on quality of data

- Think about what might have gone wrong, and how it might be revealed

- Consider the order in apply procedures

---
class: indent

# Acknowledgments

Alan Attie<br/>
Gary Churchill<br/>
Dan Gatti<br/>
Alexandra Lobo<br/>
Federico Rey<br/>
&#346;aunak Sen<br/>
Lindsay Traeger<br/>
Brian Yandell



---
class: middle, indent

# &nbsp;

Slides: [`bit.ly/2018ctc`](https://bit.ly/2018ctc)
.cc0-button[ [![CC0](CC0_button.svg)](https://creativecommons.org/publicdomain/zero/1.0) ]

[`kbroman.org`](https://kbroman.org)

[`github.com/kbroman`](https://github.com/kbroman)

[`@kwbroman`](https://twitter.com/kwbroman)
