---
title: "Cleaning genotype data for</br>diversity outbred mice"
author: "[Karl Broman](https://kbroman.org)"
date: 'Biostatistics & Medical Informatics<br/>University of Wisconsin&ndash;Madison<br/><a href="https://kbroman.org"><code>kbroman.org</code></a><br/><a href="https://github.com/kbroman"><code>github.com/kbroman</code></a><br/><a href="https://twitter.com/kwbroman"><code>@kwbroman</code></a>'
output:
    xaringan::moon_reader:
        lib_dir: libs
        nature:
            highlightStyle: github
            highlightLines: true
            countIncrementalSlides: false
            slideNumberFormat: '%current%'
            ratio: "16:9"
        css: [default, 'kbroman.css']
        chakra: libs/remark-latest.min.js
---

```{r setup, include=FALSE}
ratio <- "16:9"
fig_height <- ifelse(ratio=='4:3', 9, 7.5)
curves_height <- ifelse(ratio=='4:3', 580, 540)
curves_width <- ifelse(ratio=='4:3', 800, 1080)
options(htmltools.dir.version = FALSE)
knitr::opts_chunk$set(dev="svg", echo=FALSE,
                      fig.height=fig_height,
                      fig.width=15)
xaringan::summon_remark() # download latest version, locally
suppressMessages(library(fst))
library(qtl2)
library(broman)
```

## Heterogeneous stock

```{r dofig}
source("R/hs_fig.R")
```

---

## Diversity outbred mouse data

- Collaboration with Alan Attie, Gary Churchill, Brian Yandell,
  Josh Coon, Federico Rey, and many others

- 500 DO mice

- GigaMUGA SNP arrays

- RNA-seq data on pancreatic islets

- Microbiome data (16S and shotgun sequencing)

- protein and lipid measurements by mass spec


---


## Principles

What might have gone wrong?

How could it be revealed?



---

## Possible problems

- Sample duplicates

- Sample mix-ups

- Bad samples

- Bad markers

- Genotyping errors in founders


---

## Sex swaps

```{r load_data}
x <- read.fst("../Data/attieDO_chrXint.fst")
y <- read.fst("../Data/attieDO_chrYint.fst")
do <- readRDS("../Data/attieDO_v0.rds")

# for samples in replicate, figure out which one to keep
tab <- table(do$covar$mouse)
dup_mice <- names(tab[tab>1])
dup_mice <- sprintf("DO%03d", as.numeric(sub("DO", "", dup_mice)))
nm <- n_missing(do, summary="prop")
nm_before <- nm[dup_mice]
nm_after <- nm[paste0(dup_mice, "_b")]
keep <- names(nm_before)[nm_before < nm_after]

# samples to omit
omit <- dup_mice
omit[omit %in% keep] <- paste0(omit[omit %in% keep], "_b")

x <- x[,is.na(match(names(x), omit))]
y <- y[,is.na(match(names(y), omit))]
ids <- ind_ids(do)
do <- do[ids[!(ids %in% omit)],]
```

```{r yint_vs_xint_all_probes}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
xm <- colMeans(x[,-1], na.rm=TRUE)
ym <- colMeans(y[,-1], na.rm=TRUE)
purple_green <- brocolors("web")[c("purple", "green")]
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, "All markers", adj=1, cex=1.8)
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
```

---

## Sex swaps

```{r yint_vs_xint_selected_probes}
file <- "R/_cache/xymarkers.RData"
if(file.exists(file)) {
    load(file)
} else {
    xp <- apply(x[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    yp <- apply(y[,-1], 1, function(a) -log10(t.test(a ~ do$is_female)$p.value))
    xmar <- x[xp > -log10(0.05/length(xp)),1]
    ymar <- y[yp > -log10(0.05/length(yp)),1]
    save(xmar, ymar, file=file)
}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
xm <- colMeans(x[x[,1] %in% xmar, -1], na.rm=TRUE)
ym <- colMeans(y[y[,1] %in% ymar, -1], na.rm=TRUE)
prop_missing <- n_missing(do, "ind", "prop")
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity")
mtext(side=3, adj=1, cex=1.8, "Selected markers")
to_label <- c(names(sort(xm)[1:2]), # really low X intensity
              names(xm)[xm < 0.4 & ym < 0.1], # low X and Y
              names(xm)[do$is_female & ym > 0.2]) # female but high Y intensity
label <- sub("_b$", "", to_label)
yadd <- 0.02*c(1,1,1,1,-1.7,1)
xadj <- c(0.5, 0.5, 1, 0, 0.5, 0.5)
for(i in seq_along(to_label))
    text(xm[to_label[i]], ym[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
```

---

## Sex swaps

```{r yint_vs_xint_selected_probes_size_missing}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, ym, bg=purple_green[do$is_female + 1],
         xlab="ave X chr intensity",
         ylab="ave Y chr intensity", cex=1-prop_missing)
mtext(side=3, adj=1, cex=1.8, "Selected markers")
to_label <- c(names(sort(xm)[1:2]), # really low X intensity
              names(xm)[xm < 0.4 & ym < 0.1], # low X and Y
              names(xm)[do$is_female & ym > 0.2]) # female but high Y intensity
label <- sub("_b$", "", to_label)
yadd <- 0.02*c(1,1,1,1,-1.7,1)
xadj <- c(0.5, 0.5, 1, 0, 0.5, 0.5)
for(i in seq_along(to_label))
    text(xm[to_label[i]], ym[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
```

---

## Sex swaps

```{r het_vs_xint}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
prop_het <- rowSums(do$geno$X == 2)/rowSums(do$geno$X != 0)
grayplot(xm, prop_het, bg=purple_green[do$is_female + 1],
         ylab="proportion het on X",
         xlab="average X chr intensity")
on_top <- (do$is_female & xm < 0.4)
points(xm[on_top], prop_het[on_top], pch=21, bg=purple_green[2])
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
to_label <- c(names(xm)[xm < 0.4 & prop_het > 0.2],
              names(xm)[do$is_female & xm > 0.35 & xm < 0.4])
xadd <- 0.003*c(1,1,1,1,-2.7)
yadd <- 0.035*c(0,0,0,0,1)
xadj <- c(0,0,0,0,0.5)
label <- sub("_b$", "", to_label)
for(i in seq_along(to_label))
    text(xm[to_label[i]]+xadd[i], prop_het[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
```


---

## Sex swaps

```{r het_vs_xint_size_missing}
par(mar=c(6.1,5.1,1.6,1.1), cex=1.7, cex.axis=1.3, cex.lab=1.3)
grayplot(xm, prop_het, bg=purple_green[do$is_female + 1],
         ylab="proportion het on X",
         xlab="average X chr intensity",
         cex=1-prop_missing)
on_top <- (do$is_female & xm < 0.4)
points(xm[on_top], prop_het[on_top], pch=21, bg=purple_green[2],
         cex=1-prop_missing[on_top])
legend("topright", pch=21, pt.bg=purple_green, c("male", "female"), bg="gray90")
to_label <- c(names(xm)[xm < 0.4 & prop_het > 0.2],
              names(xm)[do$is_female & xm > 0.35 & xm < 0.4])
xadd <- 0.003*c(1,1,1,1,-2.7)
yadd <- 0.035*c(0,0,0,0,1)
xadj <- c(0,0,0,0,0.5)
label <- sub("_b$", "", to_label)
for(i in seq_along(to_label))
    text(xm[to_label[i]]+xadd[i], prop_het[to_label[i]]+yadd[i], label[i],
         adj=c(xadj[i], 0.5))
```

---

## Sample duplicates

(histogram of proportion matching genotypes)

---

## Sample mix-ups

(general illustration of the idea)

---

## Sample mix-ups: RNA-seq phenotypes

---

## Microbiome data

---

## Sample mix-ups: Microbiome data


---

## Sample quality

- Proportion missing data

- Genotype frequencies

- No. crossovers

- No. apparent genotyping errors

---

## Marker quality

- Proportion missing data

- Genotype frequencies

- No. apparent genotyping errors


---

## Founder genotyping errors



---

## Summary

- Quality of results depends on quality of data

- Think about what might have gone wrong, and how it might be revealed

- Consider the order in apply procedures
