<!DOCTYPE html>
<html>
  <head>
    <title>Cleaning genotype data for diversity outbred mice</title>
    <meta charset="utf-8">
    <meta name="author" content="Karl Broman" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <script src="libs/kePrint/kePrint.js"></script>
    <link rel="stylesheet" href="kbroman.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Cleaning genotype data for </br>diversity outbred mice
### <a href="https://kbroman.org">Karl Broman</a>
### Biostatistics &amp; Medical Informatics<br/>University of Wisconsinâ€“Madison<br/><br/><a href="https://kbroman.org"><code>kbroman.org</code></a><br/><a href="https://github.com/kbroman"><code>github.com/kbroman</code></a><br/><a href="https://twitter.com/kwbroman"><code><span class="citation">@kwbroman</span></code></a><br/>Slides: <a href="https://bit.ly/2018ctc"><code>bit.ly/2018ctc</code></a>

---




# Heterogeneous stock

![](index_files/figure-html/dofig-1.svg)&lt;!-- --&gt;

---

# Diversity outbred mouse data

- 500 DO mice

- GigaMUGA SNP arrays (114k SNPs)

- RNA-seq data on pancreatic islets

- Microbiome data (16S and shotgun sequencing)

- protein and lipid measurements by mass spec

- Collaboration with Alan Attie, Gary Churchill, Brian Yandell,
  Josh Coon, Federico Rey, and many others

---


# Principles

What might have gone wrong?

How could it be revealed?

--

Also, just make a bunch of graphs.

--

If you see something weird, try to figure it out.



---

# Possible problems

- Sample duplicates

- Sample mix-ups

- Bad samples

- Bad markers

- Genotyping errors in founders


---
class: inverse, middle, center

# What to look at first?


---

# Missing data per sample



![](index_files/figure-html/plot_missing-1.svg)&lt;!-- --&gt;

---

# Missing data per sample

![](index_files/figure-html/missing_genotypes_labeled-1.svg)&lt;!-- --&gt;


---
class: inverse, middle, center

# Swapped sex labels

---

# Average SNP intensity on X and Y chr


![](index_files/figure-html/yint_vs_xint_all_probes-1.svg)&lt;!-- --&gt;

---

# Average SNP intensity on X and Y chr

![](index_files/figure-html/yint_vs_xint_selected_probes-1.svg)&lt;!-- --&gt;

---

# Average SNP intensity on X and Y chr

![](index_files/figure-html/yint_vs_xint_selected_probes_labeled-1.svg)&lt;!-- --&gt;


---

# Heterozygosity vs SNP intensity on X chr

![](index_files/figure-html/het_vs_xint-1.svg)&lt;!-- --&gt;


---

# Heterozygosity vs SNP intensity on X chr

![](index_files/figure-html/het_vs_xint_labeled-1.svg)&lt;!-- --&gt;

---
class: inverse, middle, center

# Sample duplicates

---

# Percent matching genotypes between pairs

![](index_files/figure-html/prop_match-1.png)&lt;!-- --&gt;

---

# Percent matching genotypes between pairs

![](index_files/figure-html/prop_match_labeled-1.png)&lt;!-- --&gt;

---
class: inverse, middle, center

# Sample mix-ups

---

# Sample mix-ups

![](index_files/figure-html/mixup_illustration_1-1.svg)&lt;!-- --&gt;

---

# Sample mix-ups

![](index_files/figure-html/mixup_illustration_2-1.svg)&lt;!-- --&gt;

---

# Sample mix-ups

![](index_files/figure-html/mixup_illustration_3-1.svg)&lt;!-- --&gt;

---

# Sample mix-ups

![](index_files/figure-html/mixup_illustration_4-1.svg)&lt;!-- --&gt;

---

# RNA-seq sample mix-ups: distance matrix

![](index_files/figure-html/gve_dist_image-1.png)&lt;!-- --&gt;

---

# RNA-seq sample mix-ups: min vs self distance

![](index_files/figure-html/gve_best_vs_self_unlabeled-1.svg)&lt;!-- --&gt;

---

# RNA-seq sample mix-ups: min vs self distance

![](index_files/figure-html/gve_best_vs_self_labeled-1.svg)&lt;!-- --&gt;


---

# RNA-seq sample mix-ups: detail

![](index_files/figure-html/rnaseq_problems_detail-1.svg)&lt;!-- --&gt;

---

# Microbiome data


![](index_files/figure-html/microbiome_illustration-1.svg)&lt;!-- --&gt;

---

# Sample mix-ups: Microbiome data


- Impute genotypes at all SNPs in DNA samples

- Map microbiome reads to mouse genome; find reads overlapping a SNP

- For each pair of samples (DNA + microbiome):

  - Focus on reads that overlap a SNP where that DNA sample is
    homozygous

  - Distance = proportion of reads where SNP allele doesn't match DNA
    sample's genotype


---

# Microbiome DO361 vs DNA DO361

&lt;table class="table table-striped" style="font-size: 48px; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; AA &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; BB &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 939,918 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1,044 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2,998 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 125,962 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;




---

# Microbiome DO360 vs DNA DO360

&lt;table class="table table-striped" style="font-size: 48px; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; AA &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; BB &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 2,661,645 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 190,188 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 427,685 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 202,335 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

---

# Microbiome DO360 vs DNA DO370

&lt;table class="table table-striped" style="font-size: 48px; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt;   &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; AA &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; BB &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; A &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 3,137,751 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1,475 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; B &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 7,461 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 310,369 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

---

# Microbiome mix-ups: min vs self distance

![](index_files/figure-html/microbiome_best_vs_self-1.svg)&lt;!-- --&gt;

---

# Microbiome mix-ups: min vs self distance

![](index_files/figure-html/microbiome_best_vs_self_labeled-1.svg)&lt;!-- --&gt;


---
class: inverse, middle, center

# Sample quality

---

# Missing data per sample

![](index_files/figure-html/missing_genotypes_labeled-1.svg)

---

# Array intensities

![](index_files/figure-html/array_intensities_densities-1.png)&lt;!-- --&gt;


---

# Allele frequencies, by individual

![](index_files/figure-html/genotype_freq-1.svg)&lt;!-- --&gt;

---

# Allele frequencies, by individual

![](index_files/figure-html/allele_freq_subset-1.svg)&lt;!-- --&gt;

---

# Allele frequencies, by individual

![](index_files/figure-html/allele_freq_subset_wdensity-1.svg)&lt;!-- --&gt;

---

# Genotype frequencies, by individual


![](index_files/figure-html/geno_freq_byind-1.svg)&lt;!-- --&gt;

---

# Genotype frequencies, by individual

![](index_files/figure-html/geno_freq_byind_wlabels-1.svg)&lt;!-- --&gt;


---

# Heterozygosities, by individual

![](index_files/figure-html/het_byind-1.svg)&lt;!-- --&gt;


---

# Genotype probabilities (one mouse on one chr)

![](index_files/figure-html/plot_genoprob-1.png)&lt;!-- --&gt;

---

# Genome reconstruction (one mouse)

![](index_files/figure-html/plot_onegeno-1.svg)&lt;!-- --&gt;

---

# Percent missing vs number of crossovers

![](index_files/figure-html/missing_v_nxo-1.svg)&lt;!-- --&gt;



---

# Percent missing vs number of crossovers

![](index_files/figure-html/missing_v_nxo_labeled-1.svg)&lt;!-- --&gt;

---

# Percent missing vs number of crossovers

![](index_files/figure-html/missing_v_nxo_subset-1.svg)&lt;!-- --&gt;

---

# Percent missing vs number of crossovers

![](index_files/figure-html/missing_v_nxo_subset_labeled-1.svg)&lt;!-- --&gt;

---

# No. crossovers by generation

![](index_files/figure-html/nxo_by_wave-1.svg)&lt;!-- --&gt;

---

# Estimated percent of genotyping errors

![](index_files/figure-html/genotype_errors-1.svg)&lt;!-- --&gt;

---

# Estimated percent of genotyping errors

![](index_files/figure-html/genotype_errors_subset-1.svg)&lt;!-- --&gt;



---
class: inverse, middle, center

# Marker quality

---

# Proportion missing data


![](index_files/figure-html/hist_missing_bymar-1.png)&lt;!-- --&gt;

---

# Allele frequencies, by marker

![](index_files/figure-html/allele_freq_bymar-1.svg)&lt;!-- --&gt;

---


# Allele frequencies, by marker

![](index_files/figure-html/allele_freq_bymar_wdensity-1.svg)&lt;!-- --&gt;

---

# Genotype frequencies, by marker


![](index_files/figure-html/geno_freq_bymar-1.png)&lt;!-- --&gt;

---


# Heterozygosities, by marker

![](index_files/figure-html/het_bymar-1.svg)&lt;!-- --&gt;


---

# Genotyping error rates

![](index_files/figure-html/genotype_errors_bymar-1.png)&lt;!-- --&gt;

---

# Genotyping error rate vs percent missing

![](index_files/figure-html/error_vs_missing_bymar-1.png)&lt;!-- --&gt;

---

# Genotyping error rate vs percent missing

![](index_files/figure-html/error_vs_missing_bymar_logscale-1.png)&lt;!-- --&gt;

---

# Nice markers

![](index_files/figure-html/snpint_nice-1.svg)&lt;!-- --&gt;


---

# Crap markers

![](index_files/figure-html/snpint_crap-1.svg)&lt;!-- --&gt;

---

# More crap markers

![](index_files/figure-html/snpint_more_crap-1.svg)&lt;!-- --&gt;



---

# One bad blob

![](index_files/figure-html/snpint_onebadblob-1.svg)&lt;!-- --&gt;

---

# Wrong genomic coordinates

![](index_files/figure-html/snpint_wrong_location-1.svg)&lt;!-- --&gt;


---

# Puzzling no calls

![](index_files/figure-html/snpint_nocalls-1.svg)&lt;!-- --&gt;






---
class: inverse, middle, center

# Founder genotyping errors

---

# One founder missing

![](index_files/figure-html/snpint_foundermissing-1.svg)&lt;!-- --&gt;

---

# Another case with one founder missing


![](index_files/figure-html/snpint_foundermissing_again-1.svg)&lt;!-- --&gt;



---
class: compressed_bullets


# Summary

- Quality of results depends on quality of data

- Think about what might have gone wrong, and how it might be revealed

- Pulling out the bad samples is the most important thing

- Sex swaps: look at array intensities

- Look for sample duplicates, and if possible sample mix-ups

- Samples: missing data, array intensities, crossovers, number of errors

- Markers: lots of reasons for the bad ones

---
class: indent

# Acknowledgments

Alan Attie&lt;br/&gt;
Gary Churchill&lt;br/&gt;
Dan Gatti&lt;br/&gt;
Alexandra Lobo&lt;br/&gt;
Federico Rey&lt;br/&gt;
&amp;#346;aunak Sen&lt;br/&gt;
Lindsay Traeger&lt;br/&gt;
Brian Yandell



---
class: middle, indent

# &amp;nbsp;

Slides: [`bit.ly/2018ctc`](https://bit.ly/2018ctc)
.cc0-badge[ [![CC0](cc-zero.svg)](https://creativecommons.org/publicdomain/zero/1.0) ]

[`kbroman.org`](https://kbroman.org)

[`github.com/kbroman`](https://github.com/kbroman)

[`@kwbroman`](https://twitter.com/kwbroman)
    </textarea>
<script src="libs/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%",
"ratio": "4:3"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
