<!DOCTYPE html>
<html>
  <head>
    <title>Cleaning genotype data fordiversity outbred mice</title>
    <meta charset="utf-8">
    <meta name="author" content="Karl Broman" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="kbroman.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Cleaning genotype data for</br>diversity outbred mice
### <a href="https://kbroman.org">Karl Broman</a>
### Biostatistics &amp; Medical Informatics<br/>University of Wisconsinâ€“Madison<br/><a href="https://kbroman.org"><code>kbroman.org</code></a><br/><a href="https://github.com/kbroman"><code>github.com/kbroman</code></a><br/><a href="https://twitter.com/kwbroman"><code><span class="citation">@kwbroman</span></code></a>

---




## Heterogeneous stock

![](index_files/figure-html/dofig-1.svg)&lt;!-- --&gt;

---

## Diversity outbred mouse data

- Collaboration with Alan Attie, Gary Churchill, Brian Yandell,
  Josh Coon, Federico Rey, and many others

- 500 DO mice

- GigaMUGA SNP arrays

- RNA-seq data on pancreatic islets

- Microbiome data (16S and shotgun sequencing)

- protein and lipid measurements by mass spec


---


## Principles

What might have gone wrong?

How could it be revealed?



---

## Possible problems

- Sample duplicates

- Sample mix-ups

- Bad samples

- Bad markers

- Genotyping errors in founders


---

## Sex swaps



![](index_files/figure-html/yint_vs_xint_all_probes-1.svg)&lt;!-- --&gt;

---

## Sex swaps

![](index_files/figure-html/yint_vs_xint_selected_probes-1.svg)&lt;!-- --&gt;

---

## Sex swaps

![](index_files/figure-html/yint_vs_xint_selected_probes_size_missing-1.svg)&lt;!-- --&gt;

---

## Sex swaps

![](index_files/figure-html/het_vs_xint-1.svg)&lt;!-- --&gt;


---

## Sex swaps

![](index_files/figure-html/het_vs_xint_size_missing-1.svg)&lt;!-- --&gt;

---

## Sample duplicates

![](index_files/figure-html/prop_match-1.svg)&lt;!-- --&gt;

---

## Sample duplicates

![](index_files/figure-html/prop_match_labeled-1.svg)&lt;!-- --&gt;

---

## Sample mix-ups

(general illustration of the idea)

---

## Sample mix-ups: RNA-seq phenotypes

---

## Microbiome data


![](index_files/figure-html/microbiome_illustration-1.svg)&lt;!-- --&gt;

---

## Sample mix-ups: Microbiome data


---

## Sample quality

- Proportion missing data

- Genotype frequencies

- No. crossovers

- No. apparent genotyping errors

---

## Marker quality

- Proportion missing data

- Genotype frequencies
p
- No. apparent genotyping errors


---

## Founder genotyping errors



---

## Summary

- Quality of results depends on quality of data

- Think about what might have gone wrong, and how it might be revealed

- Consider the order in apply procedures
    </textarea>
<script src="libs/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%",
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
